Userland/Libraries/LibPDF/Shading.cpp

```diff
@@ -237,6 +237,247 @@ PDFErrorOr<void> AxialShading::draw(Gfx::Painter& painter, Gfx::AffineTransform
+            if (function_array->size() != static_cast<size_t>(common_entries.color_space->number_of_components()))
+                return Error::malformed_error("Function array must have as many elements as color space has components");
+            for (size_t i = 0; i < function_array->size(); ++i) {
+                auto function = TRY(Function::create(document, TRY(document->resolve_to<Object>(function_array->at(i)))));
+                if (TRY(function->evaluate(to_array({ 0.0f }))).size() != 1)
+                    return Error::malformed_error("Function must have 1 output component");
+                TRY(functions_vector.try_append(move(function)));
+            }
+            return functions_vector;
+        }
+        auto function = TRY(Function::create(document, function_object));
+        if (TRY(function->evaluate(to_array({ 0.0f }))).size() != static_cast<size_t>(common_entries.color_space->number_of_components()))
+            return Error::malformed_error("Function must have as many output components as color space");
+        return function;
+    }());
+
+    // "(Optional) An array of two boolean values specifying whether to extend the
+    //  shading beyond the starting and ending circles, respectively. Default value:
+    //  [ false false ]."
+    bool extend_start = false;
+    bool extend_end = false;
+    if (shading_dict->contains(CommonNames::Extend)) {
+        auto extend_array = TRY(shading_dict->get_array(document, CommonNames::Extend));
+        if (extend_array->size() != 2)
+            return Error::malformed_error("Extend must have 2 elements");
+        extend_start = extend_array->at(0).get<bool>();
+        extend_end = extend_array->at(1).get<bool>();
+    }
+
+    return adopt_ref(*new RadialShading(move(common_entries), start, start_radius, end, end_radius, t0, t1, move(functions), extend_start, extend_end));
+}
+
+PDFErrorOr<void> RadialShading::draw(Gfx::Painter& painter, Gfx::AffineTransform const& inverse_ctm)
+{
+    auto& bitmap = painter.target();
+
+    auto scale = painter.scale();
+    auto clip_rect = painter.clip_rect() * scale;
+
+    Vector<float, 4> color_components;
+    color_components.resize(m_common_entries.color_space->number_of_components());
+
+    // FIXME: Do something with m_common_entries.b_box if it's set.
+
+    // FIXME: Use smaller box if the circles are nested and the outer circle is not extended.
+
+    for (int y = clip_rect.top(); y < clip_rect.bottom(); ++y) {
+        for (int x = clip_rect.left(); x < clip_rect.right(); ++x) {
+            Gfx::FloatPoint point = inverse_ctm.map(Gfx::FloatPoint { x, y } / scale);
+
+            // The spec explains how to get a point given s. We want to solve the inverse problem:
+            // The current pixel is at p. We want to find the s where (c(s) - p)^2 = r(s)^2.
+            // Per spec, the circle depending on s is c(s) = c0 + s * (c1 - c0), r(s) = r0 + s * (r1 - r0).
+            // Putting in the values:
```

Comment:
nit: Slightly clarify c(s) is the center?

Reply:
Done.